src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java

```diff
@@ -13,12 +34,368 @@
+     */
+    public AIState dump()
+    {
+        if(!worker.isWorkerAtSiteWithMove(wareHouse.getLocation(), MIN_DISTANCE_TO_WAREHOUSE))
+        {
+            return DUMPING;
+        }
+
+        wareHouse.getTileEntity().dumpInventoryIntoWareHouse(worker.getInventoryCitizen());
+        gatherTarget = null;
+
+        return START_WORKING;
+    }
+
+
+    /**
+     * Deliver the items to the hut.
+     * @return the next state.
+     */
+    private AIState deliver()
+    {
+        AbstractBuildingWorker ownBuilding = getOwnBuilding();
+        if(ownBuilding instanceof BuildingDeliveryman)
+        {
+            AbstractBuilding buildingToDeliver = ((BuildingDeliveryman) ownBuilding).getBuildingToDeliver();
+            if (buildingToDeliver != null)
+            {
+                if(!worker.isWorkerAtSiteWithMove(buildingToDeliver.getLocation(), MIN_DISTANCE_TO_WAREHOUSE))
+                {
+                    return DELIVERY;
+                }
+
+                InventoryCitizen workerInventory  = worker.getInventoryCitizen();
+                for(int i = 0; i < workerInventory.getSizeInventory(); i++)
+                {
+                    ItemStack stack = workerInventory.getStackInSlot(i);
+                    if(stack == null)
+                    {
+                        continue;
+                    }
+
+                    if(InventoryUtils.addItemStackToInventory(buildingToDeliver.getTileEntity(), stack))
+                    {
+                        workerInventory.removeStackFromSlot(i);
+                    }
+                }
+                buildingToDeliver.setOnGoingDelivery(false);
+            }
+        }
+        return START_WORKING;
+    }
+
+    /**
+     * Prepare deliveryman for delivery.
+     * Check if the building still needs the item and if the required items are still in the warehouse.
+     * @return the next state to go to.
+     */
+    private AIState prepareDelivery()
+    {
+        AbstractBuildingWorker ownBuilding = getOwnBuilding();
+        if(ownBuilding instanceof BuildingDeliveryman)
+        {
+            AbstractBuilding buildingToDeliver = ((BuildingDeliveryman) ownBuilding).getBuildingToDeliver();
+            if(buildingToDeliver != null)
+            {
+                boolean ableToDeliver = wareHouse.getTileEntity().checkInWareHouse(buildingToDeliver);
+                if(!ableToDeliver)
+                {
+                    return START_WORKING;
+                }
+                itemsToDeliver = new ArrayList<>(buildingToDeliver.getNeededItems());
+                return GATHER_IN_WAREHOUSE;
+            }
+        }
+        return START_WORKING;
+    }
+
+    private AIState gatherItemsFromWareHouse()
+    {
+        AbstractBuildingWorker ownBuilding = getOwnBuilding();
+        if (ownBuilding instanceof BuildingDeliveryman)
+        {
+            AbstractBuilding buildingToDeliver = ((BuildingDeliveryman) ownBuilding).getBuildingToDeliver();
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'buildingToDeliver' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)